daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Interning.scala

```diff
@@ -0,0 +1,36 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.daml.lf.data
+
+/** Utilities for interning objects.
+  * For interning strings, use String.intern.
+  */
+object Interning {
+
+  /** Intern an object, that is, reuse a previously created identical object. */
+  trait Internable[T] {
+    val interned: java.util.Map[T, T] =
+      java.util.Collections.synchronizedMap[T, T](new java.util.WeakHashMap[T, T])
+
+    def intern(x: T): T = {
+      val y = interned.putIfAbsent(x, x)
```

Comment:
IIUC This will prevent `T`s from ever being collected. [From the Javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/WeakHashMap.html):

> The value objects in a WeakHashMap are held by ordinary strong references. Thus care should be taken to ensure that value objects do not strongly refer to their own keys, either directly or indirectly, since that will prevent the keys from being discarded.

When we had a similar problem in Ermine, @dolio (and I believe @ekmett) made [this cache](https://bitbucket.org/ermine-language/ermine-scala/src/123bcc9ea2cbd852ed9fe3e0267a48d43bd441e6/core/src/main/scala/com/clarifi/reporting/Cache.scala?at=default); it's under [BSD 2-clause](https://bitbucket.org/ermine-language/ermine-scala/src/default/COPYING), we should just use it or something like it and avoid all potential problems with weak/soft key maps. No GC magic is involved; the limit is a function of frequency of interning and how much space you devote to the cache in advance.